### PR TITLE
load_image - decode b64encode and encodebytes strings

### LINKED ISSUE #30114: Image classification pipeline fails with base64.encodebytes() input
### System Info

transformers - 4.39.1
python 3.10.11
platform - ubuntu 22.04

### Who can help?

_No response_

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

``` python
from PIL import Image
import base64
from io import BytesIO
from transformers import AutoModelForImageClassification, AutoFeatureExtractor
model = AutoModelForImageClassification.from_pretrained("microsoft/resnet-18")
tokenizer = AutoFeatureExtractor.from_pretrained("microsoft/resnet-18")
tr_model = {"model": model, "image_processor": tokenizer}
vision_model = transformers.pipeline(
    task="image-classification", model=model, image_processor=tokenizer
)

buffered = BytesIO()
im = torch.rand((256, 256,3))
image = Image.fromarray(im.numpy().astype('uint8'), 'RGB')
image.save(buffered, format="JPEG")
img_str1 = base64.b64encode(buffered.getvalue()).decode("utf-8")
pred1 = vision_model.predict(img_str1)

img_str2 = base64.encodebytes(buffered.getvalue()).decode("utf-8")
pred2 = vision_model.predict(img_str2)`
```

results in this error:
Traceback (most recent call last):
  File "/home/azureuser/workspace/AzureMlCli/test_oss.py", line 22, in <module>
    pred2 = vision_model.predict(img_str2)
  File "/anaconda/envs/vision_finetune/lib/python3.10/site-packages/transformers/pipelines/base.py", line 921, in predict
    return self(X)
  File "/anaconda/envs/vision_finetune/lib/python3.10/site-packages/transformers/pipelines/image_classification.py", line 158, in __call__
    return super().__call__(images, **kwargs)
  File "/anaconda/envs/vision_finetune/lib/python3.10/site-packages/transformers/pipelines/base.py", line 1162, in __call__
    return self.run_single(inputs, preprocess_params, forward_params, postprocess_params)
  File "/anaconda/envs/vision_finetune/lib/python3.10/site-packages/transformers/pipelines/base.py", line 1168, in run_single
    model_inputs = self.preprocess(inputs, **preprocess_params)
  File "/anaconda/envs/vision_finetune/lib/python3.10/site-packages/transformers/pipelines/image_classification.py", line 161, in preprocess
    image = load_image(image, timeout=timeout)
  File "/anaconda/envs/vision_finetune/lib/python3.10/site-packages/transformers/image_utils.py", line 326, in load_image
    raise ValueError(
ValueError: Incorrect image source. Must be a valid URL starting with `http://` or `https://`, a valid path to an image file, or a base64 encoded string. Got /9j/4AAQSkZJRgABAQAAAQABAAD/2wBDAAgGBgcGBQgHBwcJCQgKDBQNDAsLDBkSEw8UHRofHh0a
HBwgJC4nICIsIxwcKDcpLDAxNDQ0Hyc5PTgyPC4zNDL/2wBDAQkJCQwLDBgNDRgyIRwhMjIyMjIy
MjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjL/wAARCAEAAQADASIA
AhEBAxEB/8QAHwAAAQUBAQEBAQEAAAAAAAAAAAECAwQFBgcICQoL/8QAtRAAAgEDAwIEAwUFBAQA
AAF9AQIDAAQRBRIhMUEGE1FhByJxFDKBkaEII0KxwRVS0fAkM2JyggkKFhcYGRolJicoKSo0NTY3
ODk6Q0RFRkdISUpTVFVWV1hZWmNkZWZnaGlqc3R1dnd4eXqDhIWGh4iJipKTlJWWl5iZmqKjpKWm
p6ipqrKztLW2t7i5usLDxMXGx8jJytLT1NXW19jZ2uHi4+Tl5ufo6erx8vP09fb3+Pn6/8QAHwEA
AwEBAQEBAQEBAQAAAAAAAAECAwQFBgcICQoL/8QAtREAAgECBAQDBAcFBAQAAQJ3AAECAxEEBSEx
BhJBUQdhcRMiMoEIFEKRobHBCSMzUvAVYnLRChYkNOEl8RcYGRomJygpKjU2Nzg5OkNERUZHSElK
U1RVVldYWVpjZGVmZ2hpanN0dXZ3eHl6goOEhYaHiImKkpOUlZaXmJmaoqOkpaanqKmqsrO0tba3
uLm6wsPExcbHyMnK0tPU1dbX2Nna4uPk5ebn6Onq8vP09fb3+Pn6/9oADAMBAAIRAxEAPwD5/ooo
oAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiig
AooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKAC
iiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKK
KKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooo
oAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiig
AooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKAC
iiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKK
KKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooo
oAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiig
AooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKAC
iiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKK
KKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooo
oAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiig
AooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKAC
iiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKK
KKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooo
oAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiig
AooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooA//2Q==
. Failed with Non-base64 digit found

Image encoded with b64encode is successful, but encodebytes results in the above error.

### Expected behavior

Image classification pipeline works well when input is given in base64.base64encode() format but fails with base64.encodebytes() format. Shouldn't the pipeline work on both the input formats?

### PR DESCRIPTION
**Pull request by amyeroberts**
# What does this PR do?

Updates base64 decoding login to decode base64 string encoded with either `base64.b64encode` or `base64.encodebytes`

Fixes #30114

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

### CODE REVIEW

File: src/transformers/image_utils.py [hunk 1]
```diff
@@ -320,7 +320,7 @@ def load_image(image: Union[str, "PIL.Image.Image"], timeout: Optional[float] =
 
             # Try to load as base64
             try:
-                b64 = base64.b64decode(image, validate=True)
```
Comment (ydshieh) [inline src/transformers/image_utils.py:323]: Could you explain a bit why `b64decode` here can't handle the data encoded by `encodebytes`?

File: src/transformers/image_utils.py [hunk 2]
```diff
@@ -320,7 +320,7 @@ def load_image(image: Union[str, "PIL.Image.Image"], timeout: Optional[float] =
...
+                b64 = base64.decodebytes(image.encode() if isinstance(image, str) else image)
```
Comment (ydshieh) [inline src/transformers/image_utils.py:323]: So the fix is to make it handle string?

Comment (ydshieh) [reply src/transformers/image_utils.py:323]: But `image` is always `str` at this `if` branch `if isinstance(image, str):`, so I feel something is strange here.

**Comment by ydshieh** [comment]
Hi @amyeroberts  Let me know your thoughts on my comments whenever you get some bandwidth

File: src/transformers/image_utils.py (hunk 2 above)
Comment (amyeroberts) [inline src/transformers/image_utils.py:323]:
```suggestion
                b64 = base64.decodebytes(image.encode())
```

**Comment by amyeroberts** [comment]
@ydshieh OK, so the solution I put in here was a hacky one I found worked without digging too much into it - sorry for the rushed solution. 

So, `base64.decodebytes` can decode the bytes string from both `base64.b64encode` and `base64.encodebytes` and `base64.decode` can decode the string of `base64.b64encode` and `base64.encodebytes`. However, this will fail if we have `validate=True`:

```
import base64
from io import BytesIO
from PIL import Image

import torch

buffered = BytesIO()
im = torch.rand((256, 256,3))
image = Image.fromarray(im.numpy().astype('uint8'), 'RGB')
image.save(buffered, format="JPEG")

base64.b64decode(base64.encodebytes(buffered.getvalue()), validate=True)
```

This is because [encodebytes](https://docs.python.org/3/library/base64.html#base64.encodebytes) inserts newline characters into the encoded bytes string after every 76 bytes if output and ensures there's a trailing new line. This fails validation with `Error: Non-base64 digit found`

The solution to enable these images would be to remove `validate=True` from the current `b64decode` call. I'm not sure we should do this, as the bytes string from `base64.encodebytes(s)` can contain arbitrary binary data. WDYT?

File: src/transformers/image_utils.py (hunk 2 above)
Comment (ydshieh) [reply src/transformers/image_utils.py:323]: I still see we have 

> image.encode() if isinstance(image, str) else image

??

**Comment by ydshieh** [comment]
> The solution to enable these images would be to remove validate=True from the current b64decode call. I'm not sure we should do this, as the bytes string from base64.encodebytes(s) can contain arbitrary binary data. WDYT?

This is a good question. I tried to feed some bytes string without. Without adding extra arguments to `load_image`, not sure if we can keep an option about the check. Do think think it is worth a new argument?

**Comment by ydshieh** [comment]
I tried to decode some arbitrary bytes and `decodebytes` could still fail. It will succeed only if the bytes is a base64 encoded data.
So I think we can safely use `decodebytes` as you do in this PR.

```python
import base64


s = b'data to be encoded'
print(type(s))

data = base64.decodebytes(s)
print(data)
```

**Comment by amyeroberts** [comment]
> So I think we can safely use decodebytes as you do in this PR.

Awesome. Thanks for looking into this! I'll tidy up the code and let you know when it's ready for re-review.

**Comment by amyeroberts** [comment]
@ydshieh Ready for re-review 🤗

**Review by ydshieh — APPROVED** [review]
Thank you for the fix, the iterations and the explanations!

**Pull request merged**